lib/ruby_lsp/requests/support/rails_document_client.rb

```diff
@@ -0,0 +1,117 @@
+# typed: strict
+# frozen_string_literal: true
+
+require "net/http"
+
+module RubyLsp
+  module Requests
+    module Support
+      class RailsDocumentClient
+        RAILS_DOC_HOST = "https://api.rubyonrails.org"
+        SUPPORTED_RAILS_DOC_NAMESPACES = T.let(
+          Regexp.union(
+            /ActionDispatch/, /ActionController/, /AbstractController/, /ActiveRecord/, /ActiveModel/, /ActiveStorage/,
+            /ActionText/, /ActiveJob/
+          ).freeze,
+          Regexp,
+        )
+
+        RAILTIES_VERSION = T.let(
+          [*::Gem::Specification.default_stubs, *::Gem::Specification.stubs].find do |s|
+            s.name == "railties"
+          end&.version&.to_s, T.nilable(String)
+        )
+
+        class << self
+          extend T::Sig
+          sig { returns(T.nilable(T::Hash[String, T::Array[T::Hash[Symbol, String]]])) }
+          def rails_documents
+            @rails_documents ||= T.let(begin
+              return unless RAILTIES_VERSION
+
+              $stderr.puts "Fetching Rails Documents..."
+              # If the version's doc is not found, e.g. Rails main, it'll be redirected
+              # In this case, we just fetch the latest doc
+              response = if Gem::Version.new(RAILTIES_VERSION).prerelease?
+                Net::HTTP.get_response(URI("#{RAILS_DOC_HOST}/js/search_index.js"))
+              else
+                Net::HTTP.get_response(URI("#{RAILS_DOC_HOST}/v#{RAILTIES_VERSION}/js/search_index.js"))
+              end
+
+              if response.code == "200"
+                process_search_index(response.body)
+              else
+                $stderr.puts("Response failed: #{response.inspect}")
+                nil
+              end
+            rescue StandardError => e
+              $stderr.puts("Exception occurred when fetching Rails document index: #{e.inspect}")
+            end, T.nilable(T::Hash[String, T::Array[T::Hash[Symbol, String]]]))
```

Comment:
I still think this huge `begin`/`end` block inside a `T.let` is very unreadable, and this method mixes up the building of the rails documents with memoization of it.

So, I will still suggest that the building is split out to a specific builder method.